### DISCUSSION TITLE
Best agile rate periods

### DISCUSSION THREAD

**Discussion by mattsgreen** [BottlecapDave/HomeAssistant-OctopusEnergy#306] [Q&A]
**Describe the feature**

The Octopus Compare app has a great feature where it provides a list of best periods for Agile pricing for 1hr, 2hr, 3hr, 4hr, 5hr, 6hr.

Example:
![Screenshot_20230710-215025](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/6796359/4d101aae-e647-4c8e-a3e8-ef3e99c8e9b3)

**Expected behavior**

I'd really like to be able to present the same information in a HA dashboard to support the family with best times to consume energy. I'd also like to be able to do a morning notification for when would be the best time to put the washing machine on for example based on it.

I think the logic for this is already in the Target Rates sensors - but I can't work out how (if at all) I could get the times for them to present it as above?

**Comment by BottlecapDave** [comment]
Hi there. I've moved this to a discussion as this falls outside of integration features.

You will need a target rate sensor for each hour blocks you want to display. For example

```
Name: agile_1_hour
Hours: 1
From: 16:00
To: 16:00
```

You will then use [templating sensors](https://www.home-assistant.io/integrations/template/#state-based-template-binary-sensors-buttons-images-numbers-selects-and-sensors) to expose the attributes from the sensors. To get the information you're after you will have the following templates (you might need to adjust to get the format you desire)

Time:
```
{{ state_attr('binary_sensor.octopus_energy_agile_1_hour', 'target_times')[0]['valid_from'] }}
```

Time To
```
{{ state_attr('binary_sensor.octopus_energy_agile_1_hour', 'target_times')[0]['valid_from'] - now() }}
```

Average Price
```
{{ state_attr('binary_sensor.octopus_energy_agile_1_hour', 'overall_average_cost') }}
```

You can then set these sensors in HA entity cards to expose the information.

If you're feeling more adventurous, you could add some custom cards that support templating out of the box, like [mushroom cards](https://github.com/piitaya/lovelace-mushroom/blob/main/docs/cards/template.md) or write your own custom card like the [energy rates card](https://github.com/lozzd/octopus-energy-rates-card)

**Reply by mattsgreen** [reply]
That's fantastic! Thank you - I'll give it a try!

**Reply by mattsgreen** [reply]
Seems to be working well! 
Out of interest, why do you suggest using the first 'valid_from' entry, rather than the 'next_time' attribute?

**Reply by BottlecapDave** [reply]
I suggested the times as it won't change until the next 14 hour period comes around. Whereas next_time will potentially be undefined once the discovered cheap period has passed. But really it's whatever works best for you :)
